test/CodeGenILTests/IndexingMicroTests.cs

```diff
@@ -0,0 +1,217 @@
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            Assert.AreEqual(4, output.Values.ToList()[3][0]);
+        }
+
+        [Test]
+        [Category("Failure")]//to accomplish this either need to generate a function call, or emit conditional il.
+        public void IndexIntoRange_WithNegativeIndex_Ident()
+        {
+            var dscode = @"
+import(""DesignScriptBuiltin.dll"");
+a = [0,1,2,3,4,5,6,7,8,9,10,11,12];
+b = -5;
+c=a[b];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            Assert.AreEqual(8, output.Values.ToList()[2][0]);
+        }
+        [Test]
+        [Category("Failure")]//I can't understand this failure, the IL is valid but the loading the element retrieves an IList instead of a string...
+        public void IndexIntoArray_StringArray()
+        {
+            var dscode = @"
+import(""DesignScriptBuiltin.dll"");
+a = [""A"",""B"",""C""];
+b=a[2];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            Assert.AreEqual("C", output.Values.ToList()[1][0]);
+        }
+        [Test]
+        public void IndexIntoArray_ObjectArray()
+        {
+            var dscode = @"
+import(""DesignScriptBuiltin.dll"");
+import(""FFITarget.dll"");
+
+a = [FFITarget.DummyPoint2D.ByCoordinates(0,0),FFITarget.DummyPoint2D.ByCoordinates(1,1),FFITarget.DummyPoint2D.ByCoordinates(2,2)];
+b = 2;
+c=a[b];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            //TODO unclear why this index result has an extra level of nesting in the output dictionary. Replication maybe?
+            Assert.AreEqual(2, (output.Values.ToList()[2][0] as dynamic)[0].X);
+        }
+        #endregion
+        #region IList
+        [Test]
+        [Category("Failure")]//this is failing because range expression is wrapped in an extra level of nesting since it the result
+        //of a function call - 
+        public void IndexIntoRange_WithIdent()
+        {
+            var dscode = @"
+import(""DesignScriptBuiltin.dll"");
+a = 0..100..10; //a is an ILIST because of replication?
+b = 5;
+c=a[b];";
+            var ast = ParserUtils.Parse(dscode).Body;
+            var output = codeGen.EmitAndExecute(ast);
+            Assert.IsNotEmpty(output);
+            Assert.AreEqual(40, output.Values.ToList()[2][0]);
+        }
+        #endregion
+        #region dictionary
+        [Test]
+        [Category("Failure")]//this is failing because key and val are single items and dictionaries are initialized with lists of keys and vals.
+        //once that is resolved it will fail because we don't have enough type info (wrapped by function call logic IList)
```

Comment:
I'm not sure I understand this. You're saying the AST comes in as
``` 
a = Dictionary.ByKeysValues(var_keys, var_values); // var_values is a list of a single key, ...
b = var_key; // var_key is a string
c = Dictionary.ValueAtKey(a, b);
```
Assuming we know `a` is a dictionary, and we can find out the type of `b` to be a string, why can't we emit a function call to `Dictionary.ValueAtKey`. Am I missing something?

Reply:
I understand that currently we don't have a way to know the type of `a`, but what's the other issue?

Reply:
`Dictionary.ByKeysValues(var_keys, var_values);` expects 2 arrays, and when called with a single key/val pair, it throws. I guess the old VM promoted it to array.